base/ca/src/main/java/com/netscape/cms/servlet/cert/scep/CRSEnrollment.java

```diff
@@ -1244,29 +1258,45 @@ public void unwrapPKCS10(CRSPKIMessage req, CryptoContext cx)
         KeyWrapper kw;
         Cipher cip;
         EncryptionAlgorithm ea;
+        boolean padding = false;
 
         // Unwrap the session key with the Cert server key
         try {
             kw = cx.getKeyWrapper();
+            AlgorithmParameterSpec keyWrapConfig = null;
+            if(mUseOAEPKeyWrap) {
+                keyWrapConfig = new OAEPParameterSpec(OAEP_SHA, "MGF1", MGF1ParameterSpec.SHA256, PSource.PSpecified.DEFAULT);
+                padding = true;
+            }
+            kw.initUnwrap(cx.getPrivateKey(), keyWrapConfig);
 
-            kw.initUnwrap(cx.getPrivateKey(), null);
+            switch(String.valueOf(mEncryptionAlgorithm)) {
+                case "DES3":
+                    skt = SymmetricKey.DES3;
+                    ea = EncryptionAlgorithm.DES3_CBC;
+                    break;
+                case "AES":
+                    skt = SymmetricKey.AES;
+                    ea = EncryptionAlgorithm.AES_128_CBC;
+                    break;
+                default:
+                    skt = SymmetricKey.DES;
+                    ea = EncryptionAlgorithm.DES_CBC;
 
-            skt = SymmetricKey.Type.DES;
-            ea = EncryptionAlgorithm.DES_CBC;
-            if (mEncryptionAlgorithm != null && mEncryptionAlgorithm.equals("DES3")) {
-                skt = SymmetricKey.Type.DES3;
-                ea = EncryptionAlgorithm.DES3_CBC;
             }
 
             sk = kw.unwrapSymmetric(req.getWrappedKey(),
                               skt,
                               SymmetricKey.Usage.DECRYPT,
-                              0); // keylength is ignored
+                              padding ? ea.getKeyStrength() / 8 : 0);
 
-            skinternal = cx.getDESKeyGenerator().clone(sk);
+            if(mUseOAEPKeyWrap) {
```

Comment:
> ... if oep flag is set, we are attempting a clone operation. Not sure if that would work under the most strict fips environment. 

Actually, the clone operation is performed only if oaep flag is not set which should imply we are not in FIPS mode. In my tests the clone does not work in FIPS mode (both in Fedora and RHEL).

We can move the check inside the method and decide to perform wrap and unwrap with RSA_OAEP in all cases (FIPS and no-FIPS) but I was considering the clone to be more efficient when possible. If this is not the case I will modify accordingly.

Reply:
Well, I understand that FIPS and OAEP might go together at this point, but who knows if NSS will make a rule that cloning is not allowed in the future. It seems that things get MORE restrictive over time instead of less restrictive. Therefore it would be an easy fix to modify the  moveSymmetricKey method to check for OAEP, and simply use the standard KeyWrap.RSA instead of KeyWrap.RSA_OAEP> At some point in the future, maybe OAEP is the only thing supported, so we can make that change then. There are some examples in the KRA where we use keywrap and choose between OAEP or just regular RSA. That should suffice.